common/routing/route.go

```diff
@@ -0,0 +1,140 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package routing provides utilities to define a number of [Route] instances, which can be...
+//
+//  1. Used by servers to...
+//
+//     1.a. Register with [github.com/gorilla/mux.Router] instances via the [Route.Representation] method.
+//
+//     1.b. Deserialize HTTP path variables into a struct with the [Route.Deserialize] method.
+//
+//  2. Used by clients to construct HTTP paths for requests by calling the [Route.Path] method.
+package routing
+
+import (
+	"strings"
+)
+
+// Route represents a series of HTTP path components.
+type Route[T any] struct {
+	components []Component[T]
+}
+
+// Component represents a single HTTP path component, either a constant slug or a variable parameter.
+type Component[T any] interface {
+	// Representation is the string representation of the component for usage in a path definition, e.g. "v1" for a
+	// constant slug or "{namespace}" for a variable. This should be compatible with the format specified in
+	// the [github.com/gorilla/mux] package.
+	Representation() string
+	// Serialize returns the actual value of the slug when used in an HTTP path, e.g. "v1" for a constant slug or
+	// "test-namespace" for a variable.
```

Comment:
Probably shouldn't mention slugs in the docstring.

Reply:
Removed.